File: tutorials/tour-of-restate-typescript/src/part2/user_session.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

import * as restate from "@restatedev/restate-sdk";
import { ticketServiceApi } from "./ticket_service";
import { checkoutApi } from "./checkout";

export const userSessionRouter = restate.keyedRouter({
  // <start_add_ticket>
  async addTicket(ctx: restate.KeyedContext, userId: string, ticketId: string){
    const reservationSuccess = await ctx.rpc(ticketServiceApi).reserve(ticketId);

    if (reservationSuccess) {
      //highlight-next-line
      const tickets = (await ctx.get<string[]>("tickets")) ?? [];
      tickets.push(ticketId);
      //highlight-next-line
      ctx.set("tickets", tickets);

      ctx.sendDelayed(userSessionApi, 15 * 60 * 1000).expireTicket(userId, ticketId);
    }

    return reservationSuccess;
  },
  // <end_add_ticket>

  // <start_expire_ticket>
  async expireTicket(ctx: restate.KeyedContext, userId: string, ticketId: string){
    const tickets = (await ctx.get<string[]>("tickets")) ?? [];

    const ticketIndex = tickets.findIndex((ticket) => ticket === ticketId);

    if (ticketIndex != -1) {
      tickets.splice(ticketIndex, 1);
      ctx.set("tickets", tickets);

      ctx.send(ticketServiceApi).unreserve(ticketId);
    }
  },
  // <end_expire_ticket>

  // <start_checkout>
  async checkout(ctx: restate.KeyedContext, userId: string){
    //highlight-next-line
    const tickets = (await ctx.get<string[]>("tickets")) ?? [];

    //highlight-start
    if (tickets.length === 0) {
      return false;
    }
    //highlight-end

    const checkoutSuccess = await ctx.rpc(checkoutApi)
        .handle({ userId: userId, tickets: tickets });

    if (checkoutSuccess) {
      ctx.clear("tickets");
    }

    return checkoutSuccess;
  },
  // <end_checkout>
});

export const userSessionApi: restate.ServiceApi<typeof userSessionRouter> = {
  path: "UserSession",
};

File: tutorials/tour-of-restate-java/src/main/java/dev/restate/tour/part3/UserSession.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package dev.restate.tour.part3;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.protobuf.BoolValue;
import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.serde.jackson.JacksonSerdes;
import dev.restate.tour.generated.*;
import dev.restate.tour.generated.Tour.*;

import static dev.restate.tour.generated.CheckoutRestate.CheckoutRestateClient;
import static dev.restate.tour.generated.TicketServiceRestate.TicketServiceRestateClient;

import java.time.Duration;
import java.util.HashSet;
import java.util.Set;

public class UserSession extends UserSessionRestate.UserSessionRestateImplBase {

    // <start_add_ticket>
    // At the top of the class, define the state key: supply a name and (de)serializer
    //highlight-start
    public static final StateKey<Set<String>> STATE_KEY = StateKey.of("tickets", JacksonSerdes.of(new TypeReference<>() {}));
    //highlight-end


    @Override
    public BoolValue addTicket(ObjectContext ctx, ReserveTicket request) throws TerminalException {
        Ticket ticket = Ticket.newBuilder().setTicketId(request.getTicketId()).build();
        TicketServiceRestateClient ticketClnt = TicketServiceRestate.newClient(ctx);
        BoolValue reservationSuccess = ticketClnt.reserve(ticket).await() ;

        if (reservationSuccess.getValue()) {
            //highlight-next-line
            Set<String> tickets = ctx.get(STATE_KEY).orElseGet(HashSet::new);
            tickets.add(request.getTicketId());
            //highlight-next-line
            ctx.set(STATE_KEY, tickets);

            ExpireTicketRequest expirationRequest =
                    ExpireTicketRequest.newBuilder().setTicketId(request.getTicketId()).setUserId(request.getUserId()).build();
            UserSessionRestate.UserSessionRestateClient userSessionClnt = UserSessionRestate.newClient(ctx);
            userSessionClnt.delayed(Duration.ofMinutes(15)).expireTicket(expirationRequest);
        }

        return reservationSuccess;
    }
    // <end_add_ticket>

    // <start_expire_ticket>
    @Override
    public void expireTicket(ObjectContext ctx, ExpireTicketRequest request) throws TerminalException {
        Set<String> tickets = ctx.get(STATE_KEY).orElseGet(HashSet::new);

        boolean removed = tickets.removeIf(s -> s.equals(request.getTicketId()));

        if (removed) {
            ctx.set(STATE_KEY, tickets);
            TicketServiceRestateClient ticketClnt = TicketServiceRestate.newClient(ctx);
            ticketClnt.oneWay().unreserve(Ticket.newBuilder().setTicketId(request.getTicketId()).build());
        }
    }
    // <end_expire_ticket>

    // <start_checkout>
    @Override
    public BoolValue checkout(ObjectContext ctx, CheckoutRequest request) throws TerminalException {
        //highlight-next-line
        Set<String> tickets = ctx.get(STATE_KEY).orElseGet(HashSet::new);

        //highlight-start
        if (tickets.isEmpty()) {
            return BoolValue.of(false);
        }
        //highlight-end

        CheckoutFlowRequest checkoutFlowRequest = CheckoutFlowRequest.newBuilder().setUserId(request.getUserId()).addTickets("456").build();
        CheckoutRestateClient checkoutClnt = CheckoutRestate.newClient(ctx);
        BoolValue checkoutSuccess = checkoutClnt.checkout(checkoutFlowRequest).await();

        if (checkoutSuccess.getValue()) {
            //highlight-next-line
            ctx.clear(STATE_KEY);
        }

        return checkoutSuccess;
    }
    // <end_checkout>

}
